[RFC 0137] Nix language versioning (#137)
NixOS/rfcs: rfcs/0137-nix-language-version.md

---
feature: nix-language-version
start-date: 2022-12-12
author: @fricklerhandwerk
co-authors: @thufschmitt @Ericson2314
shepherd-team: 
shepherd-leader: 
related-issues: https://github.com/NixOS/nix/issues/7255
---

# Summary
[summary]: #summary

Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.

# Motivation
[motivation]: #motivation

Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
This proposal is first step towards overcoming that limitation.

# Detailed design
[design]: #detailed-design

Introduce a magic comment in the first line of a Nix file:

    #? Nix <version>

> where `<version>` is a released version of Nix the given file is intended to work with.

Reviewer: Should it be an exact version like `2.12.0` or should bounds like `>= 2.12.0` or `2.12.*` also be allowed?

Or is the intention for this to be an arbitrary text string? If so this sounds a lot like the [User Agent](https://en.wikipedia.org/wiki/User_agent) war, I think it should definitely be a very well-defined field here. I think basing this on a single increasing number like `builtins.langVersion` would be best, this also paves the way towards a Nix language standard.

Author: Could you elaborate on what you mean by User Agent war? How does it relate to providing information about the grammar used in a given file?

In any case, this RFC explicitly states twice that semantics are out of scope. The question purely if we want to encode that information or not, and only superficially how to do that. Everything else is left to define in future iterations.

Reviewer 2: I imagine that it should be an exact version. I'm thinking of Rust "editions" as a model. If you specify `Nix 2.12` any new breaking feature are not enabled when parsing the module. When you feel like it you update to the new syntax and update your tag line to `Nix 2.14`. 

That being said I think it probably makes more sense to learn from the Rust approach, it has worked out really well. Maybe instead of Nix versions we should just occasionally cut a new "edition" with a batch of breaking changes. Of course old editions are still supported in all new Nix builds (at least for a very long time).

Reviewer: Regarding User Agent: [wiki](https://en.wikipedia.org/wiki/User_agent#User_agent_spoofing) or [Chrome phasing out User Agent support](https://news.ycombinator.com/item?id=22685018) or some [history of User Agent](https://webaim.org/blog/user-agent-string-history/)

> In any case, this RFC explicitly states twice that semantics are out of scope.

Then I guess I don't agree with that. I don't think it make sense to introduce new syntax without defining the semantics of it.

Author: Alright, I concede I'm just reluctant to let the RFC scope spiral out of control. I'll add all semantics proposals, but please be gentle with me.

> where `<version>` is a released version of Nix the given file is intended to work with.

Reviewer: What should the default assumed value be? There's some options:
- Hardcode the default to the current Nix version `2.12.0`, this way we always know what the field is if not set, and if you use newer Nix language features you need to opt-in to use them. But this then begs the question, how do we migrate to newer defaults? This would lead to requiring the field to be mandatory at some point.
- The Nix version currently evaluating the file, but this has the problem that if a user writes a file with newer Nix version and somebody with an older Nix version tries to evaluate it, it fails without any good error message

Or should the field be mandatory with a warning (and one of the above defaults), or an error? If so, Nix could then also analyze the file to figure out which versions it supports and suggest using the oldest one of those. This then probably only works well with syntactic changes though, not semantic ones. It's also very complex to implement, doesn't really allow deprecating things at all.

Reviewer: Alternatively, if the Nix version was given in the `flake.nix` instead, it would be much more realistic to always require setting the field, since there's flake templates that can be evolved. Also a flake sets the version for the entire Flake, not just a single Nix file, so there's no big overhead to changing/assigning it. Or we could even combine the two: Allow setting the default Nix (language) version in the `flake.nix`, but allow overriding it in individual Nix files

Reviewer 2: I think the current version should be assumed. At some point we can consider a warning if you are on a "very old" version (including if you haven't specified one).

Author: ```suggestion
where `<version>` is a released version of Nix the given file is intended to work with.

If no magic comment is present, Nix 2.12 – the stable release at the time this RFC was accepted – is assumed.
```

Sounds reasonable.

> where `<version>` is a released version of Nix the given file is intended to work with.

Author: ```suggestion
where `<version>` is a released version of Nix the given file is intended to work with.

Add an appropriate parameter to commands that allow encoding the same information where where files are not involved.
```

Reviewer 3: Maybe one could say that some evaluation contexts (such as a flake with specified language version) could also be used to set the grammar version.

Author: See prior art: it doesn't make much sense to do it in a flake, because flakes are written in the same language we're trying to version.

Reviewer 2: The problem with command-line flags is that they affect everything in the evaluation. They affect all imported modules, local or remote. This means that you need a "flag day" of changing the world which is a bad idea. By having a scoped opt-in libraries, or even individual files in a library can opt-in and convert themselves one-by-one which allows gradual ecosystem migration.

Reviewer 2: I don't think you can really avoid semantics. If you are sticking a comment at the top of a file it is just an ordinary comment unless you attach semantics. If you want to achieve something with that comment you need to specify what it means.

```nix
#? Nix 2.12
"nothing"
```

# Arguments
[advantages]: #advantages

* (+) Makes explicit what can be expected to work.
* (+) Enables communicating language changes systematically.
* (+) Backwards-compatible
  * (+) Allows for gradual adoption: opt-in until semantics is implemented in Nix *and* the first backwards-incompatible change to the language is introduced.
* (+) Visually unintrusive
* (+) Self-describing and human-readable
* (+) Follows a well-known convention of using [magic numbers in files](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)
* (-) May make the appearance that changing the language is harmless.
  * (+) The convention itself is harmless and independent of the development culture around the language.
* (-) The syntax of the magic comment is arbitrary.
* (-) There is a chance of abusing the magic comment for more metadata in the future. Let's avoid that.
* (-) At least one form of comment is forever bound to begin with `#` to maintain compatibility.
* (-) It requires significant additional effort to implement and maintain an appropriate system to make use of the version information.

# Alternatives
[alternatives]: #alternatives

> - Never introduce backwards-incompatible changes to the language.

Reviewer 3: Then there is a sub-approach: keep things technically in the language but forbid their use in some codebases

Author: I'd be happy to incorporate a prepared suggestion, complete with arguments.

Reviewer 3: I think I cannot resolve the conversation in favour of a suggestio below…

---
feature: nix-language-version
start-date: 2022-12-12
author: @fricklerhandwerk
co-authors: @thufschmitt @Ericson2314
shepherd-team: 
shepherd-leader: 
related-issues: https://github.com/NixOS/nix/issues/7255
---

# Summary
[summary]: #summary

> Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.

Reviewer 3: A Nix file means that during the same evaluation different files can have different grammar versions, right?

Reviewer 2: That is my interpretation. This would allow incremental evolution.

Reviewer 4: IMO if it's declared per file then it's also meant per file. The alternative would be to have some special file in a project's root or something (like Rust and many other programming language projects do), but given Nix's rather YOLO attitude on imports I don't think that's feasible.

Whether projects want to enforce some version consistency across files however is a different question.

      * (+) We can add a command line option such that it is not more effort than `nix --version`.
        * (-) Requires adding another command line option to the public API.
  * (+) The Nix language version is decoupled Nix version numbering.
    * (+) It changes less often than the Nix version.
      * (-) That was probably due to making changes being so hard.
    * (-) There are two version numbers to keep track of.
  * (-) The magic comment should reflect that it's specifying the *Nix language* version, which would make it longer.
    * (-) Renaming the Nix language will be impossible once the mechanism is part of stable Nix.

[`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957

- Use a magic string that is incompatible with evaluators prior to the feature, e.g. `%? Nix <version>`.

  * (+) Makes clear that the file is not intended to be used without explicit handling of compatibility.
    * (-) Cannot be introduced gradually.
  * (+) Such a breaking change could also be reserved for later iterations of the Nix language.

# Prior art

- [Rust `edition` field]

>   Rust has an easier problem to solve. Cargo files are written in TOML, so the `edition` information does not have to be part of Rust itself.

Reviewer 2: I think what is missing is exactly how this value will be used. For example I think the Rust edition is an excellent reference for how this feature should behave.

1. New versions of Nix will support **multiple** grammar versions. The one used will be based on the comment at the top of the file.
2. Files with different grammar versions are **fully interoperable**.
3. Old grammar versions will continue be supported for a **long time**.

This allows evolving the grammar by allowing users to opt-in to new versions in an incremental basis. The ecosystem can gradually move one file at a time.

Author: I agree this is how it should be done, but whether this is actually possible (in terms of effort), desired, or meaningful should be decided by language maintainers. Therefore I won't add it to this RFC (yet).

---
feature: nix-language-version
start-date: 2022-12-12
author: @fricklerhandwerk
co-authors: @thufschmitt @Ericson2314
shepherd-team: 
shepherd-leader: 
related-issues: https://github.com/NixOS/nix/issues/7255
---

# Summary
[summary]: #summary

Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.

# Motivation
[motivation]: #motivation

> Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.

Reviewer: I'm answering @tazjin's [comment](https://github.com/NixOS/rfcs/pull/137#issuecomment-1355226412) in this thread

Here's some things I'd like to see deprecated/changed, but there's probably more: 
- The `with` expression since it's very ambiguous, see also https://github.com/NixOS/rfcs/pull/110
- `toString`, it does too many things, especially the list-to-string coercion is problematic
- Similarly, string coercion in derivations, which uses the same mechanism as `toString` underneath
- The `let { body` syntax
- Improving confusing path semantics, see https://github.com/NixOS/nix/issues/7338
- `__functor` and `__toString`, probably
- Making builtins more consistent, e.g. not exposing `map`, `removeAttrs`, `fetchGit` and a bunch more on the global scope

Reviewer 5: A lot of those are sensible (and I'd probably add a few more things, esp. weird scoping related behaviour/bugs), but that's a limited set and it doesn't justify (to me) reducing the barrier to making a language change by implicitly accepting them as a common occurence.

As all of these are more like cleanup of existing misbehaviours, it seems more reasonable to me to start warning about those behaviours (as tvix already does for some of them!) and then plan a future cutoff with a bump of `builtins.langVersion` that disable them.

Reviewer: Oh yeah I haven't listed any of the new features I'd like to see:
- Proper destructuring support for attribute sets (also in `let` bindings)
- Being able to reassign variable names while destructuring (something like `{ "a" = first, "b" = second }: first + second`)
- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
- Using `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
- Doing something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
- Proper support for importing and exporting Nix code, some builtin module system, especially also to enable better auto-completion (big discussion)
- More builtins in general (debatable whether that's part of the language, `builtins.langVersion` at least wasn't intended for `bulitins` additions, since you can just check with `builtins ? foo` whether a specific one is available)

While yes, every list of things we'd like to see changed is finite and might eventually be implemented, being a software engineer gives me the intuition that there will always be more things we'll need to change. Though I'd love it if we at some point just achieved endstate Nix which doesn't need any changes anymore and is perfect the way it is.

Reviewer 5: > I'd love it if we at some point just achieved endstate Nix

I think if we just clean up some of the very rough edges in existing Nix we're basically there. It's got all the crucial bits already, we can soften some edges and then declare it done. I don't have any expectation of anything turning out _perfect_ though, so maybe I'm setting a lower bar ;)

Reviewer 6: @infinisil Some of the language changes you mention cannot be handled on a per-file basis (i.e. using the tags described in this RFC). In particular, changes to `builtins` cannot be per-file since `builtins` can be passed around as an argument.

Reviewer 2: Could different versions not get different `builtin` objects when required? This is technically breaking if you try to compare the `builtins` from two different files with different versions but I don't think that is a major issue. And in that case you can avoid updating your declared version until this issue is resolved so it won't break existing code (but does have a slight chance of making updating your eval version a breaking change)

Reviewer 3: > In particular, changes to builtins cannot be per-file since `builtins` can be passed around as an argument.

If I pass builtins around, I already can augment them before passing, though.

> This proposal is first step towards overcoming that limitation.

Reviewer 7: Limitations can be good and I would argue that the Nix language's stability requirements is a good limitation. The RFC needs to provide a better motivation than “allow backwards-compatible changes” – that backwards-incompatible changes are great and good by themselves is far from obvious.

Specifically the stability of Nix's language has been praised on multiple occasions, e.g. [Nix and legacy enterprise software development: an unlikely match made in heaven](https://www.youtube.com/watch?v=0dVxtNlD5N4) – evaluating a historical nixpkgs may continue to be possible.

# Detailed design
[design]: #detailed-design

> Introduce a magic comment in the first line of a Nix file:

Reviewer 8: Good, I was about to suggest so.ething similar.

If we swap `builtins.edition`, for `builtins.languageVersion`, then this RFC could focus on building up the explicit and implicit norms that such a new practices would require. On both sides, the users of that `assert` idiom and the providers of `langVersion`.

The decisions of the latter will drive the dynamics of (good or bad) adoption.

And hence, as others have noted, the challenge in auch an RFC would be to guide the ecosystem in a direction of  responsible use:

That is, sensible balance the two atagonists _Evolution_ and _Stability_ over the horizon of time.

Backwards-compatibility, unfortunately, is not a viable answer, given sufficiently large timeframes.

    #? Nix <version>

where `<version>` is a released version of Nix the given file is intended to work with.

If no magic comment is present, Nix 2.12 – the stable release at the time this RFC was accepted – is assumed.

Add an appropriate parameter to commands that allow encoding the same information where where files are not involved.

> 

Reviewer 6: This RFC should also discuss the interaction with flakes. Probably we want the version tag on the top-level `flake.nix` to be inherited by all the `.nix` files in the flake, so you don't have to specify the version for every file in the flake.

Reviewer: Previously also briefly discussed [here](https://github.com/NixOS/rfcs/pull/137#discussion_r1050784664)

  * (+) Allows for gradual adoption: opt-in until semantics is implemented in Nix *and* the first backwards-incompatible change to the language is introduced.
* (+) Visually unintrusive
* (+) Self-describing and human-readable
* (+) Follows a well-known convention of using [magic numbers in files](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)
* (-) May make the appearance that changing the language is harmless.
  * (+) The convention itself is harmless and independent of the development culture around the language.
* (-) The syntax of the magic comment is arbitrary.
* (-) There is a chance of abusing the magic comment for more metadata in the future. Let's avoid that.
* (-) At least one form of comment is forever bound to begin with `#` to maintain compatibility.
* (-) It requires significant additional effort to implement and maintain an appropriate system to make use of the version information.

# Alternatives
[alternatives]: #alternatives

- Never introduce backwards-incompatible changes to the language.

  * (+) No additional effort required.
  * (-) Requires additions to be made very carefully.
  * (-) Makes solving some well-known problems impossible.

> 

Reviewer 3: ```suggestion

- Keep the language as implemented by Nix compatible, but officially restrict the usage of undesirable features.

* (+) Matches the current practice
* (+) Maintains usability of old Nixpkgs versions (up to definable availability of fixed-output artifacts)
* (+) Does not break the codebases we do not see before making a decision, keeping Nix a dependable upstream
* (-) Strict enforcement requires extra tooling (or an optional warning/restriction in Nix)
* (-) The implementation of the features that are no longer desirable still incur complexity/maintenance cost

```

      * (+) We can add a command line option such that it is not more effort than `nix --version`.
        * (-) Requires adding another command line option to the public API.
  * (+) The Nix language version is decoupled Nix version numbering.
    * (+) It changes less often than the Nix version.
      * (-) That was probably due to making changes being so hard.
    * (-) There are two version numbers to keep track of.
  * (-) The magic comment should reflect that it's specifying the *Nix language* version, which would make it longer.
    * (-) Renaming the Nix language will be impossible once the mechanism is part of stable Nix.

[`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957

- Use a magic string that is incompatible with evaluators prior to the feature, e.g. `%? Nix <version>`.

  * (+) Makes clear that the file is not intended to be used without explicit handling of compatibility.
    * (-) Cannot be introduced gradually.
  * (+) Such a breaking change could also be reserved for later iterations of the Nix language.

> # Prior art

Reviewer 4: Stumbled upon https://github.com/hsutter/cppfront, I think it should be discussed here too. It does not do the same thing as this RFC proposes, but there is some overlap in motivation and provided features.

Reviewer 4: On an inspired day some time ago, I sketched https://md.darmstadt.ccc.de/nix2. I think it shows that one can get pretty far with syntax changes alone, without having to touch any of the evaluation semantics of Nix.

- [Rust `edition` field]

  Rust has an easier problem to solve. Cargo files are written in TOML, so the `edition` information does not have to be part of Rust itself.

- [Flakes `edition` field]

  There had been an attempt to include an `edition` field into the Flakes schema.
  It did not solve the problem of having to evaluate the Nix expression using *some* version of the grammar.

[Rust `edition` field]: https://doc.rust-lang.org/cargo/reference/manifest.html#the-edition-field
[Flakes `edition` field]: https://discourse.nixos.org/t/nix-2-8-0-released/18714/6

# Unresolved questions
[unresolved]: #unresolved-questions

- Is the proposed magic number already in use in [other file formats](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)?
- Should we allow multiple known-good versions in one line?

> 

Reviewer 9: How does this RFC impact the development of language server, formatters, parsers of the Nix language, and other tools that depend on being able to understand the Nix language?

Right now and after 20 years we don't yet have a perfect tool for any of the tasks mentioned previously, and I worry that introducing more language versions will make that task even harder

However, there is something I find very interesting about this RFC: the fact that we can finally create a Nix language version that can be parsed 100% correctly without using GNU Bison, the GLR algorithm, Flex (lexer with states), or tools/algorithms not present in more modern programming languages like Rust

Reviewer 5: > not present in more modern programming languages like Rust

What do you mean? `rnix-parser` for example has compatibility tests with C++ Nix that guarantee that ASTs are parsed identically (in the success case, there are some minor discrepancies in error cases).

Reviewer 9: Let me clarify: there is no equivalent of GNU Bison in Rust, there is no equivalent of Flex in Rust, and there is no tool that can run the GLR algorithm in Rust. The same applies to other programming languages... GNU Bison, Flex and GLR are a particularly hard combination to port beyond c++. Yes, rnix-parser, nixel, tree-sitter-nix, and others achieved this task to some extent, but I think their authors would agree this is not an easy task, even less easy to do if we had many versions, and yes you can have all the tests and fuzzing in the world, and spend uncountable hours trying to get it correct, yet it would be nice if the Nix language just were simple to port out of the box (for instance by using a simple lexer with no states and a common parsing algorithm like LR)

What I try to express with this comment is that something I find interesting in allowing the language to evolve is making the Nix language possible to parse with simple tools and algorithms, allowing ports, easily available in all (especially more recent) programming languages. 

I'm not attacking or judging any of the existing implementations, just mentioning that this RFC can potentially allow for changing the language in a way that makes the proliferation of more implementations easier, which benefits the community

Author: > we can finally create a Nix language version that can be parsed 100% correctly without using [obscure tools]

Not sure how this RFC addresses this. Right now it would give some hints about the grammar to use for parsing a given expression.

Reviewer 9: yes, this RFC does not address this, and it shouldn't

maybe we can look at my comment from the perspective that the following two tradeoffs can cancel out together:
1. having more versions of the nix language makes it harder to create tools around it
1. making the nix language as easy to port as possible (hopefully just copy-pasting the parser generator), then it would be easier for those tools to keep up with the changes

if we don't have "2" or another mechanism, then we should not overlook "1"

another mechanisms for "2" could be making the official Nix parser "open" (in the sense that it's easy to reuse in other projects), same as #134 tries to have a standalone nix-store, having a standalone nix-parse that language servers, formatters, and other tools can use would cancel out the problem in "1"

Reviewer 5: > making the nix language as easy to port as possible (hopefully just copy-pasting the parser generator)

The syntax (and parsing thereof) is definitely a big concern in the semantics of Nix, but it is nowhere near the biggest. You would have to make many extremely invasive (not just cosmetic) changes to make Nix as simple as "just copy-pasting the parser generator", and I'm not sure how much appetite for that there is.

> having a standalone nix-parse that language servers, formatters, and other tools can use

We do! It's `rnix-parser`. It parses all correct code correctly. There's literally only one known issue, and it's about code that isn't valid anyways.

Reviewer 10: > there is no tool that can run the GLR algorithm in Rust.

This is [not](https://tree-sitter.github.io/tree-sitter/creating-parsers#:~:text=Tree-sitter is based,algorithm) [true](https://github.com/tree-sitter/tree-sitter/blob/master/docs/section-3-creating-parsers.md?plain=1#L261).

Reviewer 10: > there is no tool that can run the GLR algorithm in Rust. ... GLR are a particularly hard combination to port beyond c++.

This is true only in a very misleading sense.

Tree-sitter [implements the GLR algorithm](https://github.com/tree-sitter/tree-sitter/blob/9737fcb558b72683e81286fa46457e48d518c672/docs/section-3-creating-parsers.md?plain=1#L261) and the GLR "frontend" (parse table generator) is written in Rust.  The parsing code it emits, which implements the other half of the GLR algorithm (graph-structured stack) is C code.

It very easy to emit Rust code instead.  [It has already been done](https://github.com/tree-sitter/tree-sitter/pull/638).  The tree-sitter devs have [consciously chose not to do that](https://github.com/tree-sitter/tree-sitter/issues/1670) since it would reduce the applicability of tree-sitter (many of its users want a C-only runtime).  To be fair, it's not entirely clear if Rust's safety guarantees are a big deal here.  This isn't human-written code, so it has the same safety auditing requirements as `rustc`'s own backend.

So the reason why there are no popular all-Rust implmenentations of the "backend" of the GLR algorithm has very little to do with Rust's suitability.  It's mostly a result of there being an extremely good implementation that emits C code, and that emitter being pretty easy to audit for safety (unlike hand-written C code).

               - (-) Cannot be introduced gradually
             - (+) Such a breaking change could also be reserved for later iterations of the Nix language

   - Language versioning per "project"
       - (+) This would easily allow inheriting the language version across imports (solving a lot of issues in this proposal)
       - (-) There is currently no notion of "project"
         - (-) Attempting to establish one would be a large undertaking and not immediately help solving the problem at hand
       - (-) It would require a separate language to encode project metadata such as the language version
           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
             - (-) Recursive (albeit smaller) problem of managing the additional langauge for project metadata

   </details>

1. The file extension consists of the language version followed by `.nix`.
   Example:

   ```
   default.7.nix

>    ```

Author: Didn't know that (never bothered to look at kernels because it just works). Reason enough for me to revert back to draft.

   - Language versioning per "project" in a sidecar file
       - (+) This would easily allow inheriting the language version across imports (obviating many specifications in this proposal)
       - (-) There is currently no notion of "project" in the Nix language
         - (-) Attempting to establish one would be a large undertaking and not immediately help solving the problem at hand
       - (-) Cannot be introduced gradually, particularly relevant for a large codebase like Nixpkgs
       - (-) It would require a separate language to encode project metadata such as the language version
           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
             - (-) Recursive (albeit smaller) problem of managing the additional language for project metadata

   </details>

1. The following syntax is used for declaring the language version of a Nix expression:

   ```
   version \d*;

>    ```

Reviewer: This is problematic because you'll either have to:
- Reserve `version` as a special keyword, which would breaking compatibility itself because `version` can't be used as an identifier anymore.
- Change the parser to be able to detect whether `version` is used as a keyword or an identifier. I'm not sure if that's doable, requires arbitrary look-ahead and might be very slow

Reviewer 2: It's not clear if this is restricted to the first token in the file. If so it may be acceptable. While technically `someVariable` is a valid nix expression it will not resolve in a file unless it is a builtin global. I agree that we can't make it a keyword in general, but we can likely make it a keyword in a specific context.

Reviewer: @kevincox Ah forgot about that, yeah I think that's what's specified here, in which case this should work just fine.

Reviewer 2: I would definitely like to see some more clarification and ideally an explanation of why this is backwards-compatible.

   This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).

   The syntax is open for bikeshedding.
   Alternatives should be very short and self-describing.

   <details><summary>Arguments</summary>

   - (+) Short and fairly self-descriptive
   - (+) Not an invasive change

   </details>

   <details><summary>Alternatives</summary>

   - `use v\d*;`
     - (+) Shorter
     - (-) Does not explain much
   - `Nix language version \d*`
     - (+) Says it all
     - (-) Very long

   </details>

1. The language version declaration is optional for bare Nix expressions, and can be specified with a parameter to the evaluator.

>    If no version is specified in a bare Nix expression, assume the most recent language version supported by the evaluator.

Reviewer: This directly contradicts the earlier sentence saying that

> This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).

Author: Yes, this is to ease use of the REPL and ad hoc commands. Should add that. Or is there any objection to that?

Reviewer: Can you give an example? I guess it's unclear what "bare Nix expressions" are

       - (-) Requires more characters to account for the added expressiveness
           - This may be relevant depending on where it has to be encoded
   - Use version numbers of Nix stable releases for specifying the version of the Nix language
      - (+) More obvious to see for users what the current Nix version is rather than `builtins.langVersion`
      - (-) Would tie alternative Nix language evaluators to the rest of Nix
        - (-) One can add a command line option such that it is not more effort than `nix --version`
          - (+) That requires adding another built-in to the public API
        - (-) Using a language feature requires an additional steps from users to determine the current version
            - (-) Requires adding another command line option to the public API
      - (+) The Nix language version is decoupled Nix version numbering
        - (+) It changes less often than the Nix version
          - (-) That was probably due to making changes being so hard
            - (+) The language changing slowly is a desirable property for wider adoption
        - (-) There are two version numbers to keep track of

   </details>

> 1. The language version for Nix expressions is denoted in special syntax, at the beginning of a parse unit.

Reviewer: It should be clarified what a "parse unit" is. I guess this can be a file or just text without an associated file in case of evaluating from stdin or `--arg`

Author: ```suggestion
1. The language version for Nix expressions is denoted in special syntax, at the beginning of a parse unit.
   A parse unit is any text stream, e.g. a file or string.
```

       - (-) This loses compatibility guarantees we'd get from a stricter paradigm
       - (-) Deprives expression authors of ability to be selective with evaluator versions
     - (-) Due to hashing this is often not much different from taking *any* change into account
   - Bump version when prior evaluators would fail
     - (-) Derivation hashes may change between evaluator versions if different evaluation results were allowed within one version
       - This amounts to giving up on specifying the language exhaustively using a version label
     - (+) Fewer version increments or precautions required given current development practice

   </details>

1. Whenever Nix drops support for evaluating prior language versions, a major version bump is required.

   Example: Assuming the current language version is 8, the Nix release version is 2.20, and support for language version 6 is dropped, the next Nix release must be version 3.0

   <details><summary>Arguments</summary>

   - (+) This enforces that existing code that works will not break inadvertently when upgrading Nix

   </details>

1. Semantics are preserved across file boundaries for past language versions.

>    Example: [Best-effort interoperability](#best-effort-interoperability)

Reviewer: While that section gives some examples, it's unclear to me whether we can come up with examples that would cause problems. I think the RFC should either give an extensive set of examples, or specify the full semantics of how such interoperability is achieved.

Author: To be honest, I'd rather crowd-source the pathological examples, but here's a hint at what the implementation could look like:

```suggestion
1. Semantics are preserved across file boundaries for past language versions.

   This should be fairly straightforward to implement since values passed around in the evaluator carry all the information needed to force them.
   Newer parts of the evaluator can always wrap their values in interfaces that are accepted by older parts, as far as possible.

   Example: [Best-effort interoperability](#best-effort-interoperability)
```

Author: > depending on where a function is defined it can unwittinly accept different types! 

Yes, since for given two language versions these are not the same function. I see the problem though: if you have tangled code paths like in the module system, introducing more moving parts will make reasoning about them a lot harder.

Author: > maybe a language version is too coarse and we should allow opting in to feature sets instead, with language versions merely being aliases for sets ...

Feature sets sound like a lot of cognitive and implementation overhead. Versions as aliases for feature sets sounds like only backward compatible changes. Which is of course sensible, but doesn't allow getting rid of baggage.

Author: I disagree with imposing even more cognitive overhead to regular users than already is done by this whole proposal, already on the grounds that it's hard enough for beginners as it is. I'd weigh end user and light expression author experience higher than comfort for those digging in the bowels of Nixpkgs. Not because these contributors are somehow less important, quite the opposite. But because a) doing bowel digging is a heavy investment regardless, and I suppose the worsening by additional complexity is at worst incremental b) we have to make sure that new contributors can get onboarded somewhat reliably, and we don't want to scare them away with even more moving parts.

The premise of this proposal is that the benefits of enabling controlled change to the language will outweigh the associated non-trivial costs. Agreement on adding versioning at all is, if nothing else, testament that changes to the language are desired.

Reviewer 2: I also think that we should avoid specific feature flags. I think that in general moving the language forward is the goal. Sure, you can stay on an old version for a very long time but I think it makes sense to expect that once you do move forward you update to all of the latest changes. This avoids the need for the implementation to consider every combination of features (and possibly dependent features). It also avoids the user to need to keep a map of exactly what features are enabled when editing a file, having a linear scale of version is much easier to manager IMHO.

   - Language versioning per "project" in a sidecar file
       - (+) This would easily allow inheriting the language version across imports (obviating many specifications in this proposal)
       - (-) There is currently no notion of "project" in the Nix language
         - (-) Attempting to establish one would be a large undertaking and not immediately help solving the problem at hand
       - (-) Cannot be introduced gradually, particularly relevant for a large codebase like Nixpkgs
       - (-) It would require a separate language to encode project metadata such as the language version
           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
             - (-) Recursive (albeit smaller) problem of managing the additional language for project metadata

   </details>

1. The following syntax is used for declaring the language version of a Nix expression:

   ```
   version \d*;
   ```

> 

Author: ```suggestion
   The evaluator must ignore a shebang line (starting with `#!) at the start of files, to leave room for additional tooling.
```

This is a minor restriction to leave an escape hatch for tools like https://github.com/shlevy/nix-exec that could enable quality of life improvements down the road.

Reviewer 4: I'd like to extend this to simply saying the `version` should be the first *statement* of the file. This means that the definition only relies on the AST, which allows putting arbitrary comment data in front of it. Because even without nix-exec, one can build self-executable files, and they might require multiple shebang lines.

Author: Multiple shebang lines are subject to the shebang interpreter to process. It seems too restrictive to tie the version declaration to any particular AST. It just so happens that comments in Nix also start with `#`. And for that reason the current proposal essentially says "it's this fixed string at the beginning, except shebangs".

   This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).

   The syntax is open for bikeshedding.
   Alternatives should be very short and self-describing.

   <details><summary>Arguments</summary>

   - (+) Short and fairly self-descriptive
   - (+) Not an invasive change

   </details>

   <details><summary>Alternatives</summary>

   - `use v\d*;`
     - (+) Shorter
     - (-) Does not explain much
   - `Nix language version \d*`
     - (+) Says it all
     - (-) Very long
   - `with import <language> \d*;`
     - (+) Allows for forward compatibilty hacks such as better error messages

>      - (🔥) Because why not

Author: ```suggestion
     - (-) Will likely mislead beginners to think this is has the same semantics as the original `with import ...;`
```

   </details>

> 1. The language version declaration is optional for bare Nix expressions, and can be specified with a parameter to the evaluator.
>    If no version is specified in a bare Nix expression, assume the most recent language version supported by the evaluator.
> 
>    This is to ease use of the REPL and commands taking bare expressions.

Author: ```suggestion
1. As an exception, the language version declaration is optional for Nix expressions passed directly as an argument to the evaluator and on the REPL, and can be specified with a separate parameter to the evaluator.
   If no version is specified in a bare Nix expression, assume the most recent language version supported by the evaluator.

   This is to ease use of the REPL and evaluating ad hoc expressions.
```

@infinisil does it make this clear?

```console
$ nix-instantiate --eval b.nix
2
```

### Pathological example

Usually existing code will be interacted with by calling functions.
When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.

```nix
# a.nix
version 6;
{ value }:
value + 1
```

Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:

```nix
# b.nix
version 7;
import ./a.nix { value = %5 + 7i%; }
```

```console
$ nix-instantiate --eval bnix
error: unsupported value type `complex` at built-in operator `+`
```

## Deprecation warnings

```
nix eval --json ./test.nix
warning: URL literals are deprecated (url-literal)
         please replace this with a string: "https://nixos.org"

       at test.nix:1:1:

            1| https://nixos.org
             | ^

"https://nixos.org"

warning: The following deprecated features were used:
  - url-literal (httsp://nixos.org), 1 time

  Add `--lang-warn-verbose` to show all occurrences
  Use `--lang-no-warn=url-literal` to disable this warning.
  Use `--lang-error=url-literal` to issue errors instead of warnings.
```

> ## Drawbacks

Reviewer 3: Not sure if you should call it a drawback or a risk, but there have been mistakes in implementing the existing preferences about backwards compatibility, thus it is unclear if changes will always be properly evaluated in the future. And false promises are worse than vague promises.

Author: People make mistakes, and one cannot commit to outcomes. But one can commit to process. This proposal is to define a process.

Reviewer 2: Every RFC implementation risks implementation bugs. I don't think it needs to be mentioned.

But maybe @7c6f434c is trying to highlight that supporting multiple versions concurrently adds complexity to the project and increases maintenance burden and general likelihood of bugs down the road? I think that is fair to note as this may be more relevant than in other RFCs as this is a requirement that will affect future changes, not just a one time implementation.

Author: Added this as a risk: https://github.com/NixOS/rfcs/pull/137/files#r1230919854

Reviewer 3: > Every RFC implementation risks implementation bugs. I don't think it needs to be mentioned.

Not all implementation bugs are created equal. 

Some RFCs allow for a one-time implementation. There will be bugs, they can be found and fixed, and then the solution evolves leaving the RFC behind as an important part of the history. Some RFCs declare decision-making processes that are clearly useful even with some errors on top, and if an error is noticed then the deliberations can be corrected and eventually the decision is made.

The point here is to define compatibility guarantees, and we know there will be higher error rate than in some of the projects used as points of reference, so the question of what error rate is still useful, and how to handle violations merit more detailed consideration.

> Allowing multiple language versions to coexist complicates implementation of valuators and support tooling, and makes comprehensive test coverage harder.

Author: ```suggestion
Allowing multiple language versions to coexist complicates implementation of evaluators and support tooling, and makes comprehensive test coverage harder.
All else being equal, it may increase maintenance burden and the likelihood of introducing bugs.
```

- [A changelog of Nix (language) versions](https://code.tvl.fyi/about/tvix/docs/lang-version.md), as reflected in `builtins.langVersion`
- There have been other, sometimes breaking changes to the language that have not resulted in an increment of the language version (e.g. the recent `fetchGit` changes).
- The `builtins.toJSON 1.000001` output [changed in Nix 2.12](https://github.com/NixOS/nix/issues/8259).

Possible future changes that are in discussion:

- Remove URL literals (currently implemented via experimental-features)
- Remove the old `let { body = ... }` syntax
- Disallow leading zeroes in integer literals (such as `umask = 0022`)
- Disallow `a.x or y` if `a` is not an attribute set
- Simplifying semantics of `builtins.toString` and string interpolation
- `__functor` and `__toString`, probably
- Remove `__overrides`
- [Make `builtins` more consistent](https://github.com/NixOS/nix/issues/7290), e.g. not exposing `map`, `removeAttrs`, `fetchGit` and and others in the global scope
- Fix [imprecision in string representation of floating point numbers](https://github.com/NixOS/nix/pull/6238)
- Make the `@`-pattern consistent with destructuring

> - A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`

Reviewer 3: Is there any risk that this specific change could be designed in a non-backward-compatible way?

[ 6 7 ]
```

### Expressions are not forward compatible

```nix
# a.nix
version 6;
import ./b.nix
```

```nix
# b.nix
version 7;
builtins.langVersion
```

```console
$ nix-instantiate --eval a.nix
error: unsupported Nix language version 7
```

### Best-effort interoperability

```nix
# a.nix
version 6;
{ increment }:
increment 1
```

```nix
# b.nix
version 7;
import ./a.nix { increment = x: x + 1 }
```

Since `increment` written in version 7 carries its own implementation with it, forcing it within an expression written in version 6 just works:

```console
$ nix-instantiate --eval b.nix
2
```

### Pathological example

Usually existing code will be interacted with by calling functions.
When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.

```nix
# a.nix
version 6;
{ value }:
value + 1
```

Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:

```nix
# b.nix
version 7;
import ./a.nix { value = %5 + 7i%; }
```

```console
$ nix-instantiate --eval bnix
error: unsupported value type `complex` at built-in operator `+`
```

> 

Author: ````suggestion

In the following example, version 7 *removed* the `null` primitive, such that it can no longer be used.

```nix
# a.nix
null
```

```nix
# b.nix
version 7;
import ./a.nix
```

```console
$ nix-instantiate --eval b.nix
error: unsupported value `null`
```
````

This is consistent with best-effort interoperability:
Old code keeps working on its own, and new code has to be adapted because it was not there before the breaking change.

While this requires additional effort to adopt the new language version, expression authors can always recourse to writing new code in older versions while using newer evaluators.
This in fact allows for creating compatibility wrappers as needed.

```

## Deprecation warnings

```
nix eval --json ./test.nix
warning: URL literals are deprecated (url-literal)
         please replace this with a string: "https://nixos.org"

       at test.nix:1:1:

            1| https://nixos.org
             | ^

"https://nixos.org"

warning: The following deprecated features were used:
  - url-literal (httsp://nixos.org), 1 time

  Add `--lang-warn-verbose` to show all occurrences
  Use `--lang-no-warn=url-literal` to disable this warning.
  Use `--lang-error=url-literal` to issue errors instead of warnings.
```

## Drawbacks

Allowing multiple language versions to coexist complicates implementation of evaluators and support tooling, and makes comprehensive test coverage harder.

> All else being equal, it may increase maintenance burden and the likelihood of introducing bugs.

Author: ```suggestion
All else being equal, it may increase maintenance burden and the likelihood of introducing bugs.

Providing a pathway for introducing breaking changes bears the risk of version proliferation.
We argue though that the implementation overhead incurred by the strict compatibility requirements will by itself balance that out, while at least allowing to make such a trade-off to begin with.
```

Yet, as with any software system, in order to accommodate new insights, we want to allow the Nix language to evolve.
This sometimes involves backward-incompatible ("breaking") changes that currently cannot be made without significant downstream disruption.

Therefore we propose a mechanims and policies to introduce changes to the Nix language in a controlled and deliberate manner.
It aims to avoid breaking existing code, to prevent inadvertently breaking reproducibility, and to minimise maintenance burden for implementors and users.

## Motivating examples

Incompatible changes from the past:

- [A changelog of Nix (language) versions](https://code.tvl.fyi/about/tvix/docs/lang-version.md), as reflected in `builtins.langVersion`
- There have been other, sometimes breaking changes to the language that have not resulted in an increment of the language version (e.g. the recent `fetchGit` changes).
- The `builtins.toJSON 1.000001` output [changed in Nix 2.12](https://github.com/NixOS/nix/issues/8259).

> Possbile future changes that are in discussion:

Reviewer 11: ```suggestion
Possible future changes that are in discussion:
```

Reviewer: I'd suggest not doing typo suggestions in RFC threads, since these kind of things are really not important to the end result, so it's mostly noise compared to reviewing more important things.

Reviewer 11: Alright, it was a very easy one while I was reading :).

Author: I think it's fine. Readers can assume that closed comments are incorporated in the RFC text.

Reviewer 2: I agree. Fixing typos can also make it easier for the next reader to understand. Especially in cases where the intended word isn't obvious. And I don't want to be making judgements on what typos are worth fixing. I think it is fine to highlight them all as ideally they would all be caught and corrected before the RFC is merged.

Therefore we propose a mechanims and policies to introduce changes to the Nix language in a controlled and deliberate manner.
It aims to avoid breaking existing code, to prevent inadvertently breaking reproducibility, and to minimise maintenance burden for implementors and users.

## Motivating examples

Incompatible changes from the past:

- [A changelog of Nix (language) versions](https://code.tvl.fyi/about/tvix/docs/lang-version.md), as reflected in `builtins.langVersion`
- There have been other, sometimes breaking changes to the language that have not resulted in an increment of the language version (e.g. the recent `fetchGit` changes).
- The `builtins.toJSON 1.000001` output [changed in Nix 2.12](https://github.com/NixOS/nix/issues/8259).

Possible future changes that are in discussion:

- Remove URL literals (currently implemented via experimental-features)
- Remove the old `let { body = ... }` syntax
- Disallow leading zeroes in integer literals (such as `umask = 0022`)
- Disallow `a.x or y` if `a` is not an attribute set
- Simplifying semantics of `builtins.toString` and string interpolation

> - Remove the `let { body }` syntax

Reviewer 7: Duplicate

```suggestion
```

- `__functor` and `__toString`, probably
- Remove `__overrides`
- [Make `builtins` more consistent](https://github.com/NixOS/nix/issues/7290), e.g. not exposing `map`, `removeAttrs`, `fetchGit` and and others in the global scope
- Fix [imprecision in string representation of floating point numbers](https://github.com/NixOS/nix/pull/6238)
- Make the `@`-pattern consistent with destructuring
- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
- Use `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
- Do something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
- Better support for static analysis
- [Syntax for hexadecimal numbers](https://github.com/NixOS/nix/pull/7695)

Other discussions around language changes:

- [Make path semantics less surprsing](https://github.com/NixOS/nix/issues/7338)
- [RFC 110](https://github.com/NixOS/rfcs/pull/110) proposing a alternatives to the `with` expression
- [Nix 2 – a hypothetical syntax](https://md.darmstadt.ccc.de/s/nix2)
- [Nix language changes](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d)

## Drawbacks

> Allowing multiple language versions to coexist will over time introduce a proliferation of syntax highlighters and other tooling.

Reviewer 7: It also significantly complicates any evaluator implementation and makes comprehensive test coverage (which we do not have at the moment, I'd argue) harder.

    - (-) Strict enforcement requires extra tooling that this proposal would obviate
    - (-) The implementation of the features that are no longer desirable still incur complexity and maintenance cost
        - (-) It's still not really possible to make changes to the language

- Introduce changes to the language with language extensions or feature flags
  - (-) Combinatorial explosion
      - See [Haskell language extensions] for real-world experience
  - (-) Even more maintenance overhead
  - (+) Allows gradual adoption of features
      - (-) We already have experimental feature flags as an orthogonal mechanism, with the added benefit that they don't incur support costs and can be dropped without loss

- Never make breaking changes to the language
    - (+) No additional maintenance effort required
    - (-) Blocks improvements
    - (-) Requires additions to be made very carefully
        - (-) Even incremental changes are really expensive that way
    - (-) Makes solving some well-known problems impossible

- Continue current practice
    - (-) There is no process for breaking changes
    - (-) Breaking changes are not always announced

>     - (-) There are no means of determining compatibility between expressions and evaluator versions

Reviewer 7: That is wrong, we do have `builtins.langVersion` as well as the ability to check for the existence of builtins. Changes to language parsing have (as far as I'm aware) been practically non-breaking so far. So Nix expressions can (largely) be written in a forwards compatible and even backwards compatible way.

The problem is rather how we are using that tool at the moment:

* Breaking changes to language have on occasion not been treated as such, i.e. implemented without incrementing `builtins.langVersion`. You can still check `builtins.nixVersion` in such cases, but it could cause us to have an user agent mess in case of competing Nix implementations. Neglecting to increase the `langVersion` number indicates that this has not been a formal part of the review process so far and possibly that the coverage of the Nix language test suite. The good news is that those are not fundamental, just operational problems.
* `builtins.langVersion` is not properly documented anywhere, but in NixOS/nix's commit messages. Lack of documentation is a known problem that is actively worked on as of late.
* Our understanding of the Nix language's behavior is still lacking, since some of it seems to be accidental, but still necessary. The problem with this is of course, that it becomes harder to properly judge if changes are breaking or not. This is improving at the moment, but we'd still need to consolidate this knowledge, so it can be used as a basis of judgement.

Author: > Changes to language parsing have (as far as I'm aware) been practically non-breaking so far. So Nix expressions can (largely) be written in a forwards compatible and even backwards compatible way.

That is exactly why this proposal is made. Interesting incremental changes are currently impossible because that would break everything. The proposal allows for those changes, though at substantial cost.

>     - (-) There are no means of determining compatibility between expressions and evaluator versions

Author: ```suggestion
    - (-) There are only limited means of determining compatibility between expressions and evaluator versions (`builtins.langVersion` and checking for `builtins` with `?`)
```

@sternenseemann?

Reviewer 7: ok

       - (-) Requires more characters to account for the added expressiveness
           - This may be relevant depending on where it has to be encoded
   - Use version numbers of Nix stable releases for specifying the version of the Nix language
      - (+) More obvious to see for users what the current Nix version is rather than `builtins.langVersion`
      - (-) Would tie alternative Nix language evaluators to the rest of Nix
        - (-) One can add a command line option such that it is not more effort than `nix --version`
          - (+) That requires adding another built-in to the public API
        - (-) Using a language feature requires an additional steps from users to determine the current version
            - (-) Requires adding another command line option to the public API
      - (+) The Nix language version is decoupled Nix version numbering
        - (+) It changes less often than the Nix version
          - (-) That was probably due to making changes being so hard
            - (+) The language changing slowly is a desirable property for wider adoption
        - (-) There are two version numbers to keep track of

   </details>

> 1. The language version for Nix expressions is denoted in special syntax, at the beginning of a parse unit.

Reviewer 4: ```suggestion
1. The language version for Nix expressions is denoted in special syntax, which makes up the first statement of a file.
```

Author: I'm not aware we even have a notion of "statement".

Reviewer 4: True, the statement is wrong, "top-level expression" would probably be more precise. Yes the next sentence needs to be removed, I just didn't get GitHub to render that correctly.

```suggestion
1. The language version for Nix expressions is denoted in special syntax, which makes up the first expression in the file (or in other words, the top-level node of the AST).
```

       - (+) Allows for gradual adoption: opt-in until semantics is implemented in Nix *and* the first backward-incompatible change to the language is introduced
         - (-) This will produce surprising results if the next language version preserves syntax but changes semantics (forward compatibility)
         - (-) Requires the first language version following this proposal to be syntactically incompatible with the current language to avoid forward compatibility
       - (+) Can be made self-describing and human-readable
       - (+) Follows a well-known convention of using [magic numbers in files](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)
       - (-) May make the appearance that changing the language is harmless
         - (+) The convention itself is harmless and independent of the development culture around the language
         - (-) There is a chance of abusing the magic comment for more metadata in the future
       - (-) At least one form of comment is forever bound to begin with `#` to maintain compatibility
         - Forward compatibility is undesirable anyway
       - (-) Requires support by all tooling, lose semantics otherwise

>    - Use `assert builtins.languageVersion` in the first line of the file

Reviewer 7: ```suggestion
   - Use `assert builtins.langVersion` in the first line of the file
```

           - The `edition` field was [removed from the flakes schema](https://github.com/NixOS/nix/commit/e5ea01c1a8bbd328dcc576928bf3e4271cb55399) for that reason, as it not not allow distinguishing data from metadata
           - (+) Other languages do the same (Python, Haskell, Rust, JavaScript, ...)
             - (-) Recursive (albeit smaller) problem of managing the additional language for project metadata

   </details>

1. The following syntax is used for declaring the language version of a Nix expression:

   ```
   version \d*;
   ```

   This implies that if no language version is specified in a Nix file, it is written in version 6 (the version implemented in the stable release of Nix at the time of writing this RFC).

   The syntax is open for bikeshedding.
   Alternatives should be very short and self-describing.

   <details><summary>Arguments</summary>

   - (+) Short and fairly self-descriptive
   - (+) Not an invasive change

   </details>

   <details><summary>Alternatives</summary>

   - `use v\d*;`
     - (+) Shorter
     - (-) Does not explain much
   - `Nix language version \d*`
     - (+) Says it all
     - (-) Very long
   - `with import <language> \d*;`

>      - (+) Allows for forward compatibilty hacks such as better error messages

Reviewer 3: Another benefit: less cost for «mostly compatible» implementations. E.g. syntax-highighting the version statement exactly in the beginning is annoying, here something mostly sensible is easy.

Author: Not sure how this is different from the forward compatibility hacks in general.

```suggestion
     - (+) Allows for forward compatibility hacks such as better error messages
```

Reviewer 3: Forward compatibility hacks are something to do, here doing _nothing at all_ would be good enough for many things.

Author: This syntax would only be required for declaring a minimum evaluator version >6 to parse the expression. If you don't change the expression, there is nothing to do to begin with, as according to this proposal, every evaluator would default to version 6 anyway.

This specific argument is about what happens if your expression is written for v7 but your evaluator only supports v6.

Reviewer 3: Complex version interaction rules mean that there will be eventually a lot of «v7» files that are actually compatible with both v6 and v7. Well, unless v7 ends up completely unrelated to v6 syntactically.

Author: I'm not sure if this is going in the right direction. *Any* change to the language that is not accounted for by version bumps may fail either catastrophically or subtly (which is even worse for reproducibility) when you feed newer expressions to older evaluators. This proposal alleviates exactly that risk of expressions using *seemingly* the same syntax being confused for meaning the same thing. Example: changed semantics of `fetchGit`

Reviewer 3: Not all tools are evaluators.

           - (-) Following the latest evaluator version may inadvertently break the code for older evaluators
       - (+) Don't have to look up the latest version of the Nix language when writing code
       - (+) Does not clutter the file names for what is supposed to be the latest version of the code

   </details>

1. Language versions prior to 6 are not supported.

   <details><summary>Arguments</summary>

   - (+) Does not require additional development effort
   - (+) Prior langauge versions are not fully supported by current code already, and the rest of this proposal argues to deprecate old versions in the future in order to keep the implementation manageable
   - (-) Legacy code will not get support for managing compatibility

   </details>

1. The Nix language evaluator provides a command to output the most recent Nix language version.
   This command provides options to list all Nix language versions supported by the evaluator.

   <details><summary>Arguments</summary>

   - (+) This is for convenience to determine which features are available

   </details>

> 1. `builtins.langVersion` returns the language version used for evaluating the given expression.

Reviewer 4: This needs to be made more precise w.r.t. its interaction with cross-version interoperability.

Author: Good point. What should the following evaluate to?

```
# b.nix
builtins.langVersion
```

```nix
# b.nix
version 7;
import a.nix == builtins.langVersion
```

This whole part is actually underspecified. What would be the purpose of even having `langVersion` if the version is determined for the entire expression already?

Reviewer 2: I have the same question. I can't think of a case where an expression would want to know what language **syntax** version it is being evaluated as. The version will be static based on the version declared in the file.

I think that `builtins.langVersion` is more `builtins.builtinLibVersion` and should probably preserve its current behaviour of being global. Leave the version declaration defined here to be purely for syntax changes.

This is probably also a good simplification as `builtins` can be the same object globally. The downside would be that we can never rename anything in `builtins` or change the behaviour of a function. But that seems like less useful anyways as we can always add new functions.

Author: > This is probably also a good simplification as `builtins` can be the same object globally.

By globally you mean across versions? Possibly. In any case, the idea of versioning `builtins` separately is worth considering.

Reviewer 4: If you want cross-version interoperability the way you describe below, then `builtins` indeed needs to be versioned and your example above would evaluate to `false`. And I think this is correct, because the point of doing this versioning in the first place is to make sure code of an older version evaluates exactly the same way as it always did, even in the presence of breaking language changes.

Reviewer 2: > By globally you mean across versions?

Yes, I mean across all files in a single evaluation even if they use different syntax versions.

I think it is definitely worth considering as I'm not sure it makes sense to make breaking changes in `builtins` anyways. 

1. `builtins` can be handled by runtime logic such as an assert or a fallback. It doesn't have the same strictness as syntax changes do where the code won't run at all if the parse fails.
2. We only really need to do this if we want to make breaking changes to `builtins` such as changing the behaviour of an existing function. I'm not sure we want to do that anyways as it is likely to cause bugs for people who didn't notice the change, whereas breaking syntax changes will cause errors at parse time even if the code isn't run. (Unless we make changes that affect the behaviour of old code without causing it to fail parsing which is also something that I think we should try to avoid for this reason).

So that means that the only real thing that matching `builtins` to the syntax version would allow is removal of deprecated functions. But we would still need to keep them in Nix for older versions, so the only real benefit is that it can help people avoid bugs by pushing them off of "bad" functions. But the downside is that this push is only caught at runtime, not compile time.

The downsides would be that we aren't exposing new functions to older syntax versions, forcing an artificial update to these users (or do we add new functions all the way back but only remove/change based on the syntax flag?)

TL;DR since changes to `builtins` can be handled at runtime and can only be detected at runtime I don't think we necessarily want to tie them to the same version flag.

Reviewer 4: > I'm not sure it makes sense to make breaking changes in builtins anyways.

Counter point: I'd love to see `toString` fixed one day.

Reviewer 2: I would too, but I'm not sure that I would like to see it changed throughout my project without me knowing. I'd probably prefer something like `toString2`. Then in a decade we can remove `toString` and in another decade alias `toString2` as `toString`.

Maybe this could be handled better if we treat `builtins` as magic. So you can't do `x = builtins` and accesses to `builtins.toString` are checked as a syntax error. (We can also allow things like `builtins.newFunc ? fallback`). That way we could remove the old `toString` at the same time we add a new `toString2` or `toStr` and know that users need to make the explicit update. That way as soon as you say `version 8;` you will get a syntax error that you are using `builtins.toString` (with a helpful error message about how it is removed and you should consider `toStr` instead and here is a document describing the differences).

I guess we could publish a migration guide and expect every user to read it carefully before updating the version statement in their code, but I don't have much faith that will happen. I think we would need reliable auto-migration or errors on the new version for changes to make the process work well. (Except for possibly exception circumstances where it is worth the risk of breakage).

Reviewer 2: I doubt that any amount of documentation will avoid people introducing bugs during a version upgrade. If a compatibility change doesn't raise an error reliably people will miss it and it will bite them in the future.

For example in Rust editions you can see that all off the changes cause compile errors in previous versions. You can update the edition you are using, then make changes until your code compiles, and your code will work afterwards. I think this is a very good workflow and we should aim towards it. Expecting people to carefully read and understand the changelog and then audit their code is a recipe for bugs. I think that is best avoided.

I think we can do things like error in one version, then accept in the next, and tell people that they shouldn't skip versions when upgrading. But I think that anything that goes from working -> working with changed behaviour is very risky.

Author: Resolved, removed `builtins.langVersion` from the proposal as it's not related to the goals of this RFC.

   <details><summary>Arguments</summary>

   - (+) Could make use of it for generating Nix expressions programmatically and annotating them with the correct version
   - (+) `builtins.langVersion` is already part of the stable interface, this way we can make it pure
   - (-) Requires maintaining more API surface without a clear use case

   </details>

   <details><summary>Alternatives</summary>

   - Return the latest language version instead

      - (-) Doesn't help to determine what is used for evaluating the current expression
          - (+) Might provide opportunities for forward-compatible Nix code
          - (-) Brittle and defeats the purpose of this RFC
      - (-) Impure, the value depends on the environment

   - Don't expose `builtins.langVersion` at all

      - (+) No need to have this if the Nix files are versioned
      - (+) `builtins.langVersion` is not documented and not widely used

>       - (-) Requires a Nix language version bump to implement this RFC

Reviewer 7: This is required either way.

   </details>

> 1. Each time the language specification (currently as embodied by the Nix language evaluator) is changed, the language version must be incremented.

Reviewer 4: This is very imprecise IMO, but I already disagree with its spirit. I think that a bump in language version is needed if and only if a change is not backwards compatible: If existing nix code runs and evaluates the same way as before with that language change, bumping the version is not necessary. (I can formalize this for syntax and semantic if you want to)

The implication is that the language version feature will *not* be a replacement for "min nix version" tracking, since an old evaluator might not be able to deal with newer Nix code. This is in line with what other languages do, for example Rust has both MSRV and edition.

Author: > If existing nix code runs and evaluates the same way as before with that language change, bumping the version is not necessary.

I'm thinking in terms of such examples:

1. Nix evaluator added `builtins.foo` in Nix 2.99, a backward compatible change requiring no language version increment
2. Expression author uses `builtins.foo ? workaround` to keep their code working with Nix < 2.99
3. Expression author decides to drop support for Nix < 2.99 and now uses `builtins.foo` directly
4. Expression users running Nix < 2.99 get random breakages

This would be alleviated by instead requiring a version bump even on pure additions, as users would get an error message in place saying that their dependency needs a newer evaluator to run.

> The implication is that the language version feature will not be a replacement for "min nix version" tracking, since an old evaluator might not be able to deal with newer Nix code. 

Not sure what that means. By "language version feature" do you mean the output of `builtins.langVersion`, as opposed to declaring that the evaluator must support at least a given language version with `version <number>`? If so, as in the other comment, I'm not sure if `langVersion` should even continue to exist.

For reference: https://github.com/rust-lang/api-guidelines/discussions/231

Reviewer 4: > Nix evaluator added builtins.foo in Nix 2.99, a backward compatible change requiring no language version increment

According to my (admittedly strict) definition above, this is not a backward compatible change, since existing code `builtins.foo ? workaround` from your example evaluates to a different result with it. So additions of new `builtins` usually require a version bump, while addition of e.g. new language operators does not.

> By "language version feature" do you mean the output of builtins.langVersion, as opposed to declaring that the evaluator must support at least a given language version with version <number>?

No, I was mainly referring to the `version` statement.

Author: > additions of new `builtins` usually require a version bump, while addition of e.g. new language operators does not.

How then would you even discern that a new operator was added? The same example applies. If adding an operator to the evaluator does not bump the version, the expression author can just use it without changing the version declaration, and that will break the expression for users using of a prior version of the evaluator.

Reviewer 4: Yes, that's why I said:

> The implication is that the language version feature will *not* be a replacement for "min nix version" tracking, since an old evaluator might not be able to deal with newer Nix code. This is in line with what other languages do, for example Rust has both MSRV and edition.

Author: I still don't know what that means though. Can you expand?

Reviewer 4: Keeping with the Rust example, new syntax features may be added at any time to the language, as long as they don't break old code. Of course when using those new syntax features, older Rust tooling may choke on that. That's why projects declare a "minimum supported Rust version", MSRV.

When making changes to the language the are *not* backwards compatible in that they might change the evaluation semantics of existing code (or more likely, make existing code not compile anymore), then this is done through the "Rust edition" declaration.

MSRV and Rust edition are two distinct features. The motivation here is that most of the language changes are pure additions which only require updating the toolchain, while breaking changes are bundled up into an edition bump every n years.

Reviewer 4: The obvious downside of this split is that it ties the ecosystem to the tooling implementation, or at least its versioning. If we don't want that (see https://github.com/NixOS/rfcs/pull/137#discussion_r1226426777), we could discuss some alternatives:

1. (I think what the RFC currently suggests) Always bump the Nix version on any kind of change related to the language. I think if we didn't have to tie `builtins` to the language version this could maybe work, but I remain sceptical
2. Use `major.minor` for the Nix version, with the guarantee that minor bumps only require toolchain updates. This would require the tooling to check the minor version, and bail out if it doesn't support it.

Reviewer 2: > According to my (admittedly strict) definition above, this is not a backward compatible change, since existing code builtins.foo ? workaround from your example evaluates to a different result with it. So additions of new builtins usually require a version bump, ...

I think we should reconsider this. IIUC the current stability guarantee is that `builtins` can grow new members without being considered a breaking change. I don't see any reason to roll this back. Doing `builtins.foo ? fallback` doesn't make sense if `foo` doesn't exist in any version.

Bumping the version would lock users out of the new function until they migrate to the new syntax which seems like an unnecessary burden. (Unless we keep `builtins.langVersion` and syntax version separate, in which case it doesn't hurt to increment `builtins.langVersion` as much as we want as it is just informational)

Reviewer 4: Agreed, and I was just writing about this too. I still think that builtins should generally be part of the language due to how they work, however I would make an explicit exception to the definition of a breaking change such that introducing new builtins does *not* count. This might be a bit trickier to formalize, but definitely not impossible.

Reviewer 2: I think it is pretty easy to formalize:

1. The existence and behaviour of any documented member of `builtins` is considered stable.
2. The behaviour and existence or lack there of of any other members are not considered stable and **must not** be relied upon.

I said "documented" here but we may loosen that a bit. (And in practice a member that is known to be used even if not documented will likely be treated as if it was, we just don't guarantee that).

I also don't know if we want to add other guarantees such as `bulitins.foo ? fallback` not returning an error? But I think we don't. Touch something that isn't declared and you are on your own in terms of stability. When we add new documented members we can guarantee (explicitly or implicitly) that accessing that name on old versions will return "doesn't exist" (or document whatever behaviour it had previously, for example if we documented a previously "internal" function).

Author: I gather from this, in the proposal we should explore in more detail if it's worth having a major-minor versioning and whether versioning builtins separately makes sense.

The key point why that matters is that breaking language changes require existing code to be changed to use new evaluators.

Author: @kevincox @piegamesde added an argument in favor of semver, as handling pure additions separately would simplify evaluator implementation. Since it's the only alternative without real drawbacks, it will be promoted to be part of the proposal. Will also add details on how this interacts with builtins: I think new builtins should be handled in the minor version, then the mechanism to determine compatibility stays uniform and we don't need eval-time tricks at the call site.

   <details><summary>Arguments</summary>

     - (+) The principled solution: guarantees reproducibility given a fixed language version
     - (-) Implies additional overhead in development effort:
       - Either for Nix maintainers to accommodate that practice in the release lifecycle
         - For example, one would have to batch language changes for a version bump to limit the number of increments
           - (+) This would be beneficial for alternative implementations in terms of churn and effort to keep up
       - Or implementors of alternative evaluators catching up with changes
         - (+) Specifying the language precisely via the version actually offers alternative implementations an alternative to catching up: only support a given language version

>      - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation ) development to be more independent of the rest of Nix

Author: ```suggestion
     - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation) development to be more independent of the rest of Nix
```

       This is good, since it in turn forces stronger separation of concerns and more architectural clarity
     - (-) Prohibits best-effort attempts at evaluating expressions with possibly incompatible evaluators
       - (+) With the proposed level of strictness, one doesn't have to rely on best effort but can instead be explicit

>      - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors

Reviewer 4: Rust has the policy that breaking code that shouldn't have worked in the first place is not a breaking change (by definition).

Author: Well, that depends on whether you define bugs to be part of the interface. In lack of a specification that's not the implementation itself, and given the general aspiration around Nix to keep stuff working when not deliberately changed by users, I'm inclined to do that.

Reviewer 4: Bugs are defined as any violations of earlier made contracts and guarantees. In practice, most of the time these are unsound APIs and undefined behavior. There might be other causes as well, like incorrect algorithms.

Reviewer 7: Pointer equality is arguably a bug, but alas…

Author: > [bugs are] unsound APIs

Unsound with respect to what? What is that contract or guarantee supposed to be in our case?

> [bugs are] undefined behavior

As @sternenseemann pointed out, we don't understand the language well enough to even define undefined behavior. But even then, one would have to determine if it's fine to fill in undefined gaps without bumping the language version.

> [bugs might be] incorrect algorithms

As we have seen only recently, even an obviously inconsistent serialisation of floats is something we're stuck with forever if we *really* don't want to break existing code. I don't think the argument holds up.

What I can imagine though, therefore thanks for the intellectual resistance, is that unintended changes against previous specifications can be considered a bug. In fact, the changed floating point behavior can be said to be a bug regardless of it's more sensible. If it is not said to be a bug but an intended change, it would have required a version bump.

Reviewer 4: I didn't say these are the definitions we should use, I just showed what Rust does as an example. And yes I am aware of https://xkcd.com/1172/, and that people will depend on software's most misbehaviors. One compromise would be to allow such kind of bug fixes within n months of the feature's introduction.

> Unsound with respect to what?

Soundness is a formal property of a type system, and only because of Rust's `unsafe` does this partially become a code problem. So yes I don't think this is really applicable to Nix.

Author: > One compromise would be to allow such kind of bug fixes within n months of the feature's introduction.

From experience, talking about any sort of timing in open source development really doesn't make sense. One thing I could imagine, also from experience, is something like a stability staging that takes one or two releases.

(Whether scheduling releases is meaningful or not is an independent question and does not concern this proposal. There are alternatives to cutting a release at fixed intervals.)

Author: Outdated as references to `builtins.langVersion` were removed.

       - This could be communicated with an increment in the Nix patch-level version, as is already practice

   </details>

   <details><summary>Alternatives</summary>

   - Bump version only when evaluation result on prior version evaluators would be *substantially* different
     - (+) Leaves room for judgement by developers
       - (+) Allows controlling progression of versions to some degree
     - (-) Conversely, leaves room for sneaking in breaking changes unannounced
       - (-) This loses compatibility guarantees we'd get from a stricter paradigm
       - (-) Deprives expression authors of ability to be selective with evaluator versions
     - (-) Due to hashing this is often not much different from taking *any* change into account
   - Bump version when prior evaluators would fail
     - (-) Derivation hashes may change between evaluator versions if different evaluation results were allowed within one version
       - This amounts to giving up on specifying the language exhaustively using a version label
     - (+) Fewer version increments or precautions required given current development practice

   </details>

> 1. Whenever Nix drops support for evaluating prior language versions, a major version bump is required.

Reviewer 4: I think this RFC should decide whether it wants to treat the Nix language and `nix` as coupled, or to be just about the Nix language independent of implementation. In the latter case, implementation specific things should be moved to "examples and interactions".

   - `-Wall`
   - `-Wno-`
   - `-Werror=`

   <details><summary>Arguments</summary>

   - (+) Disallowing reuse avoids dealing with change of meaning across versions
   - (+) Not considering experimental features simplifies their handling and is not required by them being exempt from compatibility guarantees.
   - (-) Maintenance burden for everyone using these old constructs, or evaluating old revisions of Nixpkgs.
   
   </details>

   <details><summary>Alternatives</summary>

   - Opt-in warnings
     - (-) Can't really make breaking changes as people won't be warned ahead of time
   - No warnings, just errors
     - (-) Doesn't offer a transition window to users
     - (+) Easier to implement

   </details>

1. A the end of the evaluation, print statistics and explanations.
   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

## Show the current Nix language version

```console
nix --language-version
7
```

```console
nix --supported-language-versions
6
7
```

> ## Version interoperability

Reviewer 4: I think we probably want interoperability, but I don't like the current proposal. Because if I'd be strict about its implementation, an evaluator would have to have separate types for values of different versions etc., or at least tag every single value with its version. This is almost guaranteed to cause an unacceptable performance and implementation complexity overhead.

I'm not sure about it yet, but one idea would be to have the evaluator always follow the semantics of the latest version. Code of older versions needs to be transformed upon import so that its old semantics are preserved. Most changes will likely be trivial syntax transformations anyways, but other times wrapping values in compatibility objects etc. might be required.

I think that this RFC should take a look at all the proposed language changes (linked in the motivation section), and for each one make sure that the interoperability story is sensible. (I'm not saying we should implement any of these in Nix, I'm just saying that this RFC should be able to deal with any of them being implemented.)

Btw I am fine with doing "best effort" interoperability, but I'd like to see more upfront declarations of which exact things are out of scope.

Author: > one idea would be to have the evaluator always follow the semantics of the latest version

That makes sense, but I wouldn't know what that would mean concretely. Do you have an example?

Reviewer 4: The goal here is that I don't want the evaluator to have to track for every value which version it "has" in order to properly evaluate it. The motivation for this is implementation complexity, but also that cross-version compatibility should not hurt performance when it is not being used. Ideally, one could then even have this translation done in a pre-processor which is external to the actual Nix interpreter.

However, this does not fully work with language changes which remove some fundamental values from the language, e.g. removing `null` or an entire primitive type. Those values probably don't have a functional equivalent in the new language, therefore interpreters must keep knowing about them and handle them correctly.

Author: > However, this does not fully work with language changes which remove some fundamental values from the language

That would indeed fall under best effort interop. If an unsupported value is passed out of old code, new code bails. This is consistent, because the new code was not there before removing the value type.

### Versioned built-ins can be passed across file boundaries

```nix
# a.nix
version 6;
builtins.langVersion
```

```nix
# b.nix
version 7;
[ (import ./a.6.nix) builtins.langVersion ]
```

```console
$ nix-instantiate --eval b.nix
[ 6 7 ]
```

### Expressions are not forward compatible

```nix
# a.nix
version 6;
import ./b.nix
```

```nix
# b.nix
version 7;
builtins.langVersion
```

```console
$ nix-instantiate --eval a.nix
error: unsupported Nix language version 7
```

### Best-effort interoperability

```nix
# a.nix
version 6;
{ increment }:
increment 1
```

```nix
# b.nix
version 7;
import ./a.nix { increment = x: x + 1 }
```

Since `increment` written in version 7 carries its own implementation with it, forcing it within an expression written in version 6 just works:

```console
$ nix-instantiate --eval b.nix
2
```

> ### Pathological example

Reviewer 7: Note that this means that only new value types can be added, but the representation can never be changed (unless the representation before and after are isomorphic). This means expressions can be versioned, but values never. The existence of thunks complicates that of course, since they are allowed versioned contents so to speak.

Normal values can never be changed with this proposal, e.g. we can't change strings from bytestrings to vectors of unicode code points or change the size of integer types.

Author: We *can* change representation values, but new language versions would have to provide the mappings, as you say. As you pointed out, due to this proposal aiming for very ambitious compatibility goals, it implies accordingly ambitious implementations.

Usually existing code will be interacted with by calling functions.
When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.

```nix
# a.nix
version 6;
{ value }:
value + 1
```

> Here we pretend that langauge version 7 introduced a new value type and syntax for complex numbers:

Reviewer 7: ```suggestion
Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:
```

- JaveScript modules
    - .cjs and .mjs extensions for commonjs/es-modules syntax variants
    - `function() { "use strict"; return 10 }`

- [Go language changes proposal](https://github.com/golang/proposal/blob/master/design/28221-go2-transitions.md#language-changes)

  The document features a discussion of multiple approaches to evolve the language without breaking existing code, and comes to the similar conclusions as our proposal, mainly:
  - One needs a way to specify the language version
    - Interestingly, Go developers decided to equate the language version with the compiler release version, despite admitting that this may be confusing
  - Keep support for code written in older language versions
  - Breaking down into features that can be enabled separately is not practical

- [Flakes `edition` field]

  There had been an attempt to include an `edition` field into the Flakes schema.
  It did not solve the problem of having to evaluate the Nix expression using *some* version of the grammar.

  [Flakes `edition` field]: https://discourse.nixos.org/t/nix-2-8-0-released/18714/6

# Future work
[future]: #future-work

> - Define rules deciding when a change to the language is appropriate, to avoid version proliferation and limit complexity of implementations.

Reviewer 4: Should already be part of this RFC IMO.

Author: ```suggestion
```

Actually this does not concern the purpose of this RFC at all.

- Define a roadmap to introduce the next language versions, for example:
  - Version 7 commits to changes that do not require additional work on Nixpkgs:
    - Introduce the version declaration, required to distinguish versions 6 and 7
    - Deprecate URL literals
    - Deprecate the `let-body` syntax
    - Drop support for leading zeroes on integers
    - Formalise change in float representation

>   - Version 8 is released at least a year after version 7, and only when version annotations according to version 7 are fully supported in Nixpkgs

Reviewer 7: Putting a timeline on version 8 makes me doubt the statement that “should remain a rare exception” reflects the authors intentions.

Author: This isn't a timeline but a lower bound, used in the example to illustrate the scale. I wouldn't know how such changes could even be done that fast. We can remove that part, no problem.

```suggestion
  - Version 8 is released only when version annotations according to version 7 are fully supported in Nixpkgs
```

     - (-) Makes more evident that the Nix language is a distinct architectural component of the Nix ecosystem
   - (+) It's currently handled that way, no change needed apart from documentation
        - See [`builtins.langVersion`] (currently undocumented)
   - (+) Simple and unambiguous
   - (+) Concise, even in the long term, since the language is supposed to change very rarely

   [`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957

   </details>

   <details><summary>Alternatives</summary>

   - [Calendar Versioning](https://calver.org/)
       - (+) Provides information on when changes happened
           - (-) This is not needed because only compatibility information is needed
       - (-) Requires a minimum amount of characters
           - This may be relevant depending on where it has to be encoded
       - (+) Restricting to only the year would force language changes to be rare
           - (+) This would allow obvious synchronisation points with Nixpkgs releases
           - (-) It may be too much policy encoded in a mechanism
   - [Semantic Versioning](https://semver.org/)

>        - (+) Can distinguish additions from other changes
>            - (-) This is not needed for our use case, since any addition to an expression will break for older evaluators even if the major version matches

Author: ```suggestion
       - (+) Can distinguish additions from other changes
           - (-) This is not needed for our use case, since any addition to an expression will break for older evaluators even if the major version matches
       - (+) Simplifies evaluator implementation: pure additions can be guarded by the minor version cleanly within a major version's evaluator code
```

           - (-) Following the latest evaluator version may inadvertently break the code for older evaluators
       - (+) Don't have to look up the latest version of the Nix language when writing code
       - (+) Does not clutter the file names for what is supposed to be the latest version of the code

   </details>

1. Language versions prior to 6 are not supported.

   <details><summary>Arguments</summary>

   - (+) Does not require additional development effort
   - (+) Prior langauge versions are not fully supported by current code already, and the rest of this proposal argues to deprecate old versions in the future in order to keep the implementation manageable
   - (-) Legacy code will not get support for managing compatibility

   </details>

1. The Nix language evaluator provides a command to output the most recent Nix language version.
   This command provides options to list all Nix language versions supported by the evaluator.

   <details><summary>Arguments</summary>

   - (+) This is for convenience to determine which features are available

   </details>

1. `builtins.langVersion` returns the language version used for evaluating the given expression.

>    <details><summary>Arguments</summary>
> 
>    - (+) Could make use of it for generating Nix expressions programmatically and annotating them with the correct version
>    - (+) `builtins.langVersion` is already part of the stable interface, this way we can make it pure
>    - (-) Requires maintaining more API surface without a clear use case
> 
>    </details>
> 
>    <details><summary>Alternatives</summary>
> 
>    - Return the latest language version instead
> 
>       - (-) Doesn't help to determine what is used for evaluating the current expression
>           - (+) Might provide opportunities for forward-compatible Nix code
>           - (-) Brittle and defeats the purpose of this RFC
>       - (-) Impure, the value depends on the environment
> 
>    - Don't expose `builtins.langVersion` at all
> 
>       - (+) No need to have this if the Nix files are versioned
>       - (+) `builtins.langVersion` is not documented and not widely used
> 
>    </details>
> 
> 1. Each time the language specification (currently as embodied by the Nix language evaluator) is changed, the language version must be incremented.
> 
>    <details><summary>Arguments</summary>
> 
>      - (+) The principled solution: guarantees reproducibility given a fixed language version
>      - (-) Implies additional overhead in development effort:
>        - Either for Nix maintainers to accommodate that practice in the release lifecycle
>          - For example, one would have to batch language changes for a version bump to limit the number of increments
>            - (+) This would be beneficial for alternative implementations in terms of churn and effort to keep up
>        - Or implementors of alternative evaluators catching up with changes
>          - (+) Specifying the language precisely via the version actually offers alternative implementations an alternative to catching up: only support a given language version
>      - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation) development to be more independent of the rest of Nix
>        This is good, since it in turn forces stronger separation of concerns and more architectural clarity
>      - (-) Prohibits best-effort attempts at evaluating expressions with possibly incompatible evaluators
>        - (+) With the proposed level of strictness, one doesn't have to rely on best effort but can instead be explicit
>      - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors
>        - This could be communicated with an increment in the Nix patch-level version, as is already practice
> 
>    </details>
> 
>    <details><summary>Alternatives</summary>
> 
>    - Bump version only when evaluation result on prior version evaluators would be *substantially* different
>      - (+) Leaves room for judgement by developers
>        - (+) Allows controlling progression of versions to some degree
>      - (-) Conversely, leaves room for sneaking in breaking changes unannounced
>        - (-) This loses compatibility guarantees we'd get from a stricter paradigm
>        - (-) Deprives expression authors of ability to be selective with evaluator versions
>      - (-) Due to hashing this is often not much different from taking *any* change into account
>    - Bump version when prior evaluators would fail
>      - (-) Derivation hashes may change between evaluator versions if different evaluation results were allowed within one version
>        - This amounts to giving up on specifying the language exhaustively using a version label
>      - (+) Fewer version increments or precautions required given current development practice
> 
>    </details>
> 

Author: Considering the discussion so far, we should probably drop `builtins.langVersion` altogether, as it's not needed with explicit version declarations. We can remove it from the proposal (or at most hint at its removal in the roadmap) since its removal would require a version bump anyway.

```suggestion
```

  The document features a discussion of multiple approaches to evolve the language without breaking existing code, and comes to the similar conclusions as our proposal, mainly:
  - One needs a way to specify the language version
    - Interestingly, Go developers decided to equate the language version with the compiler release version, despite admitting that this may be confusing
  - Keep support for code written in older language versions
  - Breaking down into features that can be enabled separately is not practical

- [Flakes `edition` field]

  There had been an attempt to include an `edition` field into the Flakes schema.
  It did not solve the problem of having to evaluate the Nix expression using *some* version of the grammar.

  [Flakes `edition` field]: https://discourse.nixos.org/t/nix-2-8-0-released/18714/6

# Future work
[future]: #future-work

- Define rules deciding when a change to the language is appropriate, to avoid version proliferation and limit complexity of implementations.
- Define a roadmap to introduce the next language versions, for example:
  - Version 7 commits to changes that do not require additional work on Nixpkgs:

>     - Introduce the version declaration, required to distinguish versions 6 and 7

Author: ```suggestion
    - Introduce the version declaration, required to distinguish versions 6 and 7
    - Remove `builtins.langVersion`, as it's not needed any more
```

https://github.com/NixOS/rfcs/pull/137/files#r1230975053

- The `builtins.toJSON 1.000001` output [changed in Nix 2.12](https://github.com/NixOS/nix/issues/8259).

Possible future changes that are in discussion:

- Remove URL literals (currently implemented via experimental-features)
- Remove the old `let { body = ... }` syntax
- Disallow leading zeroes in integer literals (such as `umask = 0022`)
- Disallow `a.x or y` if `a` is not an attribute set
- Simplifying semantics of `builtins.toString` and string interpolation
- `__functor` and `__toString`, probably
- Remove `__overrides`
- [Make `builtins` more consistent](https://github.com/NixOS/nix/issues/7290), e.g. not exposing `map`, `removeAttrs`, `fetchGit` and and others in the global scope
- Fix [imprecision in string representation of floating point numbers](https://github.com/NixOS/nix/pull/6238)
- Make the `@`-pattern consistent with destructuring
- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
- Use `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
- Do something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
- Better support for static analysis

> - [Syntax for hexadecimal numbers](https://github.com/NixOS/nix/pull/7695)

Reviewer 4: ```suggestion
- [Syntax for hexadecimal numbers](https://github.com/NixOS/nix/pull/7695)
- [Fix multiline strings](https://github.com/NixOS/nix/issues/3759)
```

   - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors
     - This could be communicated with an increment in the Nix patch-level version, as is already practice

   </details>

   <details><summary>Alternatives</summary>

   - Bump major version only when evaluation result on prior version evaluators would be *substantially* different
     - (+) Leaves room for judgement by developers
       - (+) Allows controlling progression of versions to some degree
     - (-) Conversely, leaves room for sneaking in breaking changes unannounced
       - (-) This loses compatibility guarantees we'd get from a stricter paradigm
       - (-) Deprives expression authors of ability to be selective with evaluator versions
     - (-) Due to hashing this is often not much different from taking *any* change into account

   </details>

1. Semantics are preserved across file boundaries for past language versions.

   This should be fairly straightforward to implement since values passed around in the evaluator carry all the information needed to force them.

>    Newer parts of the evaluator can always wrap their values in interfaces that are accepted by older parts, as far as possible.

Reviewer 10: Yikes, this sounds extremely heavyweight!  You mean we're going to wrap individual thunks and integers with an additional language-version parameter?

I think this is not a great idea.

Also the example isn't terribly persuasive.  If Nix were to add complex numbers, surely it would also extend the definition of addition to cover complex-float addition.  As long as old-version code isn't able to *create* complex numbers, the behavior of a 100%-old-code codebase can't change.  So there's no harm in evaluating old code with the newer version of `+` that understands complex numbers.

Reviewer 4: The idea would be to do so at parsing time, so that no additional tracking should be needed in the evaluator. Also this would only be done to the extent necessary, integers for example are unlikely to ever get hit.

I think it is fair to have some performance overhead with cross-version code, as long as code that does not make use from it does not pay a performance penalty. Something something zero-cost …

Reviewer 10: > The idea would be to do so at parsing time

That... isn't what the RFC proposes.

Have you tried actually implementing this part of the proposal?  I really think you should, before it is approved.  You are likely to be quite surprised.

Reviewer 4: I'm not the RFC author, but at least this is what I had in mind which resulted in the text above. And I agree that some significant implementation effort will have to be conducted before approving this RFC.

>    Example: [Best-effort interoperability](#best-effort-interoperability)
> 
>    <details><summary>Arguments</summary>
> 
>    - (+) This allow for incremental evolution of code bases without having to change existing code at all.
>    - (+) Expressions that are valid today will still be valid as long as a given evaluator supports the language version they are written in
>    - (+) The root of evaluation is always at a language version determined by the user, and authors of expressions in newer language versions are responsible (and made aware of the fact by the file name signaling proposed here) to interoperate with existing code
>    - (-) Passing around incompatible values (e.g. builtins or data types) between language versions can lead to surprising errors
>      - (+) We can postpone dealing with particular issues as they arise, but the general setup should support most cases on a best(-and-minimal)-effort basis
>      - (+) In any case, such breakages will only happen when adding new code, and will never break existing code
> 
>    </details>
> 
>    <details><summary>Alternatives</summary>
> 
>    - Do not support version interoperability at all
>      - (+) Avoids any unforseen issues at no cost
>      - (-) Prohibits incremental changes, as any language update will require updating all files involved
>    - Do not support passing values to functions from older language versions
>      - (-) Calling functions is the most common use case, and you can hardly do anything without it
> 
>    </details>

Reviewer 10: ```suggestion
```

Reviewer 10: > Ideally you'd express them as code suggestions, which would save me a lot of time.

   <details><summary>Arguments</summary>

   - (+) This is for convenience to determine which features are available

   </details>

1. Whenever Nix drops support for evaluating prior language versions, a major version bump is required.

   Example: Assuming the current language version is 8.0, the Nix release version is 2.20, and support for language version 6 is dropped, the next Nix release must be version 3.0

   <details><summary>Arguments</summary>

   - (+) This enforces that existing code that works will not break inadvertently when upgrading Nix

   </details>

   <details><summary>Alternatives</summary>

   - Separate Nix development from the Nix language entirely and keep it outside of the scope of this proposal
     - (-) Currently the upstream Nix language evaluator and compatibility of expressions in Nixpkgs is closely tied to the rest of Nix, and have to take that into account
       - Further separation of concerns is out of scope for this proposal

   </details>

## Deprecation warnings and errors

1. Each language construct to deprecate relative to a prior version is given a symbolic name.
   There is a way to refer to all language constructs.

>    Examples: `url-literal`, `let-body`, `int-leading-zeros`

Reviewer 10: ... *cough* `function-equality` *cough*

[ 6 null ]
```

### Expressions are not forward compatible

```nix
# a.nix
version 6.1;
import ./b.nix
```

```nix
# b.nix
version 7.0;
builtins.null
```

```console
$ nix-instantiate --eval a.nix
error: unsupported Nix language version 7.0
```

### Best-effort interoperability

```nix
# a.nix
version 6.1;
{ increment }:
increment 1
```

```nix
# b.nix
version 7.0;
import ./a.nix { increment = x: x + 1 }
```

Since `increment` written in version 7.0 carries its own implementation with it, forcing it within an expression written in version 6.1 just works:

```console
$ nix-instantiate --eval b.nix
2
```

### Pathological example

Usually existing code will be interacted with by calling functions.
When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.

```nix
# a.nix
version 6.1;
{ value }:
value + 1
```

Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:

```nix
# b.nix
version 7.0;
import ./a.nix { value = %5 + 7i%; }
```

```console
$ nix-instantiate --eval bnix
error: unsupported value type `complex` at built-in operator `+`
```

> In the following example, assume version 7.0 *removed* floating point numbers, such that they can no longer be used.

Reviewer 10: That situation is covered by "major version bump".

I really don't think we have a persuasive example of the need for this "best effort interoperability" (in addition to which I object to the name -- it's squishy and vague, which is something Nix certainly *isn't*.)

Reviewer 10: Regarding *adding* complex numbers: If Nix were to add complex numbers, surely it would also extend the definition of addition to cover complex-float addition.  As long as old-version code isn't able to *create* complex numbers, the behavior of a 100%-old-code codebase can't change.  So there's no harm in evaluating old code with the newer version of `+` that understands complex numbers.

Reviewer 10: Lastly and least-importantly, the name "best effort interoperability" is squishy and vague, which is something Nix certainly *isn't*.

Reviewer 4: > removing complex numbers should be covered by "major version bump"

This sentence makes no sense to me in this context: The example cited here already contains a major version bump.

> As long as old-version code isn't able to create complex numbers, the behavior of a 100%-old-code codebase can't change. So there's no harm in evaluating old code with the newer version of + that understands complex numbers.

And as long as old code cannot be called with new values like complex numbers.

Reviewer 10: > And as long as old code cannot be called with new values like complex numbers.

Maybe you intend to limit this "best-effort" behavior to "newly-added variants of the [enum `ValueType`](https://github.com/NixOS/nix/blob/8350f06d6c319b250d006b195415b9b6e95d8ca1/src/libexpr/value.hh#L40-L57)".  If so you should say so.  "Best effort" is very open-ended and vague.

Reviewer 10: I'd like to suggest replacing this whole section with a single sentence saying something approximately like: future RFCs which add a language feature must explain how that language feature should interact with code whose `version` declaration is too old to be aware of that feature.

I think that would be much simpler and still ensure that the problems, if any, get solved.

> ### Best-effort interoperability
> 
> ```nix
> # a.nix
> version 6.1;
> { increment }:
> increment 1
> ```
> 
> ```nix
> # b.nix
> version 7.0;
> import ./a.nix { increment = x: x + 1 }
> ```
> 
> Since `increment` written in version 7.0 carries its own implementation with it, forcing it within an expression written in version 6.1 just works:
> 
> ```console
> $ nix-instantiate --eval b.nix
> 2
> ```
> 
> ### Pathological example
> 
> Usually existing code will be interacted with by calling functions.
> When passing values from newer versions to functions from older versions of the language, interoperatbility can only be supported on a best-effort basis.
> 
> ```nix
> # a.nix
> version 6.1;
> { value }:
> value + 1
> ```
> 
> Here we pretend that language version 7 introduced a new value type and syntax for complex numbers:
> 
> ```nix
> # b.nix
> version 7.0;
> import ./a.nix { value = %5 + 7i%; }
> ```
> 
> ```console
> $ nix-instantiate --eval bnix
> error: unsupported value type `complex` at built-in operator `+`
> ```
> 
> In the following example, assume version 7.0 *removed* floating point numbers, such that they can no longer be used.
> 
> ```nix
> # a.nix
> version 6.1;
> 1.1
> ```
> 
> ```nix
> # b.nix

Reviewer 10: ```suggestion
```

Reviewer 10: > Ideally you'd express them as code suggestions, which would save me a lot of time.

> version 7.0;
> (import ./a.nix) * 2
> ```
> 
> ```console
> $ nix-instantiate --eval b.nix
> error: unsupported type `float` for multiplication
> ```
> 
> This is consistent with best-effort interoperability:
> Old code keeps working on its own, and new code has to be adapted because it was not there before the breaking change.
> 
> While this requires additional effort to adopt the new language version, expression authors can always recourse to writing new code in older versions while using newer evaluators.
> This in fact allows for creating compatibility wrappers as needed.

Reviewer 10: ```suggestion
```

---
feature: nix-language-version
start-date: 2022-12-12
author: @fricklerhandwerk
co-authors: @thufschmitt @Ericson2314
shepherd-team: 
shepherd-leader: 
related-issues: https://github.com/NixOS/nix/issues/7255
---

# Summary
[summary]: #summary

Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.

# Motivation
[motivation]: #motivation

Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.
This proposal is first step towards overcoming that limitation.

# Detailed design
[design]: #detailed-design

> Introduce a magic comment in the first line of a Nix file:

Reviewer 3: I think we had some kind of tooling for nix-scripts where first line is shebang?

Author: Please add a reference. I don't know of such a thing and wouldn't know what to search for.

Author: I see. That's indeed cursed. Does that make an argument for the RFC or can we close this thread, since it does not preclude the usage you envision?

Author: Agreed. Addressed here: https://github.com/NixOS/rfcs/pull/137#discussion_r1050809650

    #? Nix <version>

> where `<version>` is a released version of Nix the given file is intended to work with.

Reviewer 5: We're strongly against tying the version of the language to released versions of the C++ implementation. Tvix can now evaluate nixpkgs, and hnix already could, so there are at least 2 implementations of the language for which this versioning would be meaningless.

The other thing is that setting the barrier for this too low *encourages* random changes to the language, which already owes a huge amount of its complexity to changes that were made without proper discussion.

Author: Added this to the arguments.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

```nix
#? Nix 2.12
"nothing"
```

# Arguments
[advantages]: #advantages

> * (+) Encourages developing Nix to deal with changes to the Nix language in a principled manner.

Reviewer 12: How does this help with disciplined changes? I can only see that it makes experiments in production and easier as you can retire them with the next version. 

This sounds like a feature in practice but nix will need to keep all the older features. The code in its current form will suffer even more if it isn't just a weird language but also one with a few quirks being enabled and disabled with every other version.

Reviewer 4: One way I can see around this is to require that for any version change, there exists a purely syntactical transformation that upgrades some code to the next version. We would want this anyways in order to build automatic upgrade tooling. But requiring it means that the core nix code can only worry about the latest version and the backwards compatibility can be stuffed into a corner where it is not in the way most of the time.

Author: You're right. Updated the arguments accordingly.

Reviewer 7: > One way I can see around this is to require that for any version change, there exists a purely syntactical transformation that upgrades some code to the next version. We would want this anyways in order to build automatic upgrade tooling. But requiring it means that the core nix code can only worry about the latest version and the backwards compatibility can be stuffed into a corner where it is not in the way most of the time.

This is a) doubtful (there won't only be syntactic changes) and b) not feasible, as nixpkgs *needs* to be evaluatable by multiple Nix versions.

Reviewer 4: a) is a question of scope. It would be perfectly fine to decide on making that restriction for the language version feature.

b) I don't see how this is specific to my proposal compared to the entire RFC

-------

Anyways. Let's continue the discussion of my idea at https://github.com/NixOS/rfcs/pull/137#discussion_r1052409918

* (+) Backwards-compatible
  * (+) Allows for gradual adoption: opt-in until semantics is implemented in Nix *and* the first backwards-incompatible change to the language is introduced.
* (+) Visually unintrusive
* (+) Self-describing and human-readable
* (+) Follows a well-known convention of using [magic numbers in files](https://en.m.wikipedia.org/wiki/Magic_number_(programming)#In_files)
* (-) The syntax of the magic comment is arbitrary.
* (-) There is a chance of abusing the magic comment for more metadata in the future. Let's avoid that.
* (-) At least one form of comment is forever bound to begin with `#` to maintain compatibility.
* (-) It requires significant additional effort to implement and maintain an appropriate system to make use of the version information.

# Alternatives
[alternatives]: #alternatives

- Never introduce backwards-incompatible changes to the language.

  * (+) No additional effort required.
  * (-) Requires additions to be made very carefully.
  * (-) Makes solving some well-known problems impossible.

- Use the output of [`builtins.langVersion`] for specifying the version of the Nix language.

>   * (-) `builtins.langVersion` is currently only internal and undocumented.

Reviewer 5: Then lets fix this instead! We recently investigated what caused the increase of langVersion from 5 to 6, and it wasn't actually a complicated change. There were only a few of those changes so far, and we should be able to figure out what they were and document it.

Author: Added this to the arguments.

---
feature: nix-language-version
start-date: 2022-12-12
author: @fricklerhandwerk
co-authors: @thufschmitt @Ericson2314
shepherd-team: 
shepherd-leader: 
related-issues: https://github.com/NixOS/nix/issues/7255
---

# Summary
[summary]: #summary

Introduce a convention to determine which version of the Nix language grammar to use for parsing a Nix file.

# Motivation
[motivation]: #motivation

Currently it's impossible to introduce backwards-incompatible changes to the Nix language without breaking existing setups.

> This proposal is first step towards overcoming that limitation.

Reviewer 12: We would still need a way to ensure this isn't applied arbitrarily in nixpkgs. We need CI and potentially nix features to prohibit usage of this feature and eventually newer nix versions until they can be assumed "adopted". This is the same discussion that we are having regarding bumping the minimal nix version requirement in nixpkgs.

Author: I think this is the exact same underlying issue, as @tazjin suggests, making the language version explicit would allow addressing it independent of C++ Nix. I think this is covered by these arguments:
- [makes language version explicit](https://github.com/NixOS/rfcs/pull/137/files#diff-cf8ae1b6c5672f362e65afa013d72916a1559b0fc5d48e640db1fbfeb58f8223R43)
- [changing the language should be encoded in development culture](https://github.com/NixOS/rfcs/pull/137/files#diff-cf8ae1b6c5672f362e65afa013d72916a1559b0fc5d48e640db1fbfeb58f8223R49-R50)
- [encoding the language version provides compatibility information to third-party evaluators](https://github.com/NixOS/rfcs/pull/137/files#diff-cf8ae1b6c5672f362e65afa013d72916a1559b0fc5d48e640db1fbfeb58f8223R65-R67)
